### PR TITLE
[CPU] Add new attribute to control peeling

### PR DESCRIPTION
**Pull request by banach-space**
Adds a new CPU lowering config attribute, `enable_loop_peeling`,
to control whether to run the loop peeling pass before vectorisation.

Peeling is highly beneficial when using "scalable" vectorisation - it
helps to get rid of masking:
  * https://discourse.llvm.org/t/psa-scalable-auto-vec-in-linalg-without-masking/

Similarly to other Ops (most notably `linalg.matmul`), peeling is
enabled through "pre-processing". It's "off" by default and can be
enabled with:
  * --iree-llvmcpu-vector-pproc-strategy=peel

### CODE REVIEW

**Comment by banach-space** [comment]
This is primarily to enable peeling for Convolutions. This seemed like the most canonical approach to me, but folks might disagree. I'm happy to refactor if you feel that there's a better way. Thanks!

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I see why you have it this way, but it seems a bit off that we add a new pipeline just to chose a different vectorization strategy. I think it might be better to maybe update the translation info and add the vectorization stragety to that. The translation info has a filed of just dictionary attributes that could be anything. You just need to set that and look for it when you set the vector strategy. So its is a fairly localized change.

**Comment by banach-space** [comment]
> I think it might be better to maybe update the translation info and add the vectorization stragety to that.

Thanks for the suggestion, updated. Can you take another look?

I'll update the summary once I get "+1".  I want to make sure that this is what you had in mind.

>  to chose a different vectorization strategy

Just to clarify - this is "vectorization pre-processing strategy" rather than "vectorization strategy" (as in, vectorization doesn't change). Mentioning just in case people are surprised there are no changes in the vectoriser 😅

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 1]
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
...
+  if (vecPreProcStrategy == VectorPreProcStrategy::Peeling) {
+    // Enable peeling. To this end, attach extra info to the pipeline config.
+    // This will later be extracted by LLVMCPULowerExecutableTargetPass.
+    auto ctx = convOp.getContext();
+    SmallVector<NamedAttribute> attrs;
+    auto peelAttrName = StringAttr::get(ctx, "enable_peeling");
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.push_back({peelAttrName, peelAttrVal});
+    pipelineConfig = DictionaryAttr::get(ctx, attrs);
+  }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I'm confused. I thought the VectorPreProcStrategy is None for all the convolutions in `getVectorPreProcStrategy`, but we have a special logic here. Is it because we can set the strategy to peeling through CLI flags?

I can see the value about what Mahesh suggested. We decided the strategy in `KernelDispatch.cpp`, and we want to "recover" the information when adding passes to pipeline. They could conflict. It seems to be a separate issue from enabling scalable vectors for convolutions. So there are code health issue and feature support to me.

RE code health issue, I wonder if we can set the strategy on attribute in getVectorPreProcStrategy? This way we don't have floating logics everywhere to serve the need. It will need more work to make it happen cleanly, so I think it could be done in a follow-up.

RE convolution enablement: I wonder if we can retire `addConvTileAndDecomposeExpertPassPipeline`. Instead, we just use `addMultiTilingExpertPassPipeline` for convolutions as well. To achieve it, we need to add `DecomposeConvolutionToLowerDimOpsPass` to the multi-tiling pipeline, and have an additional option for `splitVectorTransfersTo`.

**Review by MaheshRavishankar — COMMENTED** [review]
Thanks @banach-space . I think this is better, but I would wait for more feedback from Hanhan and Benoit who work on this more.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 1 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Thanks for taking a look - lots of great comments 🙏🏻 

>  I thought the VectorPreProcStrategy is None for all the convolutions in getVectorPreProcStrategy, but we have a special logic here. Is it because we can set the strategy to peeling through CLI flags?

Yes. 

We have no heuristics to select a pre-processing strategy for convs. With this patch, we would allow one to be set by the user. At some point, we can also turn peeling "ON" by default for SVE. 

Btw, I always use `getVectorPreProcStrategy` to retrieve the pre-processing strategy rather than reading [clPProcStrategy](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L128-L143) directly - this way we can avoid duplicating the logic to deal with `clPProcStrategy".

> We decided the strategy in KernelDispatch.cpp, and we want to "recover" the information when adding passes to pipeline.

Yes - we need to "forward" information:
*  from [setConvRootConfig](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L2109-L2112),
*  to [LLVMCPULowerExecutableTargetPass::runOnOperation()](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L88).

I see two options for this:
1. Through a dedicated `DispatchLoweringPassPipeline` that can be used to specialise in this [switch stmt](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L114) - that's what I tried initially,
2. Through attributes - this is what the current implementation does.

I see pros and cons for both. You have much more experience with this, so I will happily follow your suggestion.

>  They could conflict

Can you expand? Not sure I know what you meant here.

> RE code health issue, I wonder if we can set the strategy on attribute in getVectorPreProcStrategy? 

I see what you mean. There's `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert`, which are virtually identical: https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L128-L139

So, if the "design" is to limit the number of lowering expert, then those two should probably be merged?

>  It will need more work to make it happen cleanly, so I think it could be done in a follow-up.

I'm happy to refactor things, but we should definitely defer that to a separate PR. I'm worried that this can lead to some surprises. I'll be under less pressure if once convs are fast for SVE :) 

>  I wonder if we can retire addConvTileAndDecomposeExpertPassPipeline. Instead, we just use addMultiTilingExpertPassPipeline for convolutions as well. 

Possibly. but looks more complex than merging `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert`:
* [addMultiTilingExpertPassPipeline](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L349-L423),
* [addConvTileAndDecomposeExpertPassPipeline](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L425-L480)

These two have diverged a fair bit.

### Summary
There's quite a lot in here, so let me summarise. Please correct me if I got sth wrong or you disagree with the proposed approach.
1. Use **attributes** rather than a **new lowering expert** for this change,
2. In a follow-up PR, merge  `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert` (this should "just work" ™️ )
3. In a follow-up PR, try merging `addConvTileAndDecomposeExpertPassPipeline` and `addMultiTilingExpertPassPipeline` (and remove `CPUConvTileAndDecomposeExpert`?). 

Basically, I'd like to land this one without 2. and 3. I'm less confident about 3., but am happy to work on both and see how it goes.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Thanks for the good summary. It's aligned with what I'm thinking. Yes, we should go with (1) -> (2) and perhaps (3) later.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 2]
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
     scalableTileFlags.emplace_back(numTilingDims, false);
   }
 
+  DictionaryAttr pipelineConfig;
+  if (vecPreProcStrategy == VectorPreProcStrategy::Peeling) {
+    // Enable peeling. To this end, attach extra info to the pipeline config.
+    // This will later be extracted by LLVMCPULowerExecutableTargetPass.
+    auto ctx = convOp.getContext();
+    SmallVector<NamedAttribute> attrs;
+    auto peelAttrName = StringAttr::get(ctx, "enable_peeling");
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Instead of having a magic string floating in different files, how about adding a `getPeelAttrName` to CPUUtils?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 3]
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
...
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.push_back({peelAttrName, peelAttrVal});
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: nit: emplace_back because we intend to call the constructor.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp [hunk 4]
```diff
@@ -140,6 +140,16 @@ void LLVMCPULowerExecutableTargetPass::runOnOperation() {
...
+    auto peelAttrName = StringAttr::get(funcOp.getContext(), "enable_peeling");
+    auto trueAttr = BoolAttr::get(funcOp.getContext(), true);
+
+    DictionaryAttr config = translationInfo.getConfiguration();
+    if (config) {
+      auto peelAttr = translationInfo.getConfiguration().getNamed(peelAttrName);
+      if (peelAttr && peelAttr->getValue() == trueAttr) {
+        pipelineOpts.enablePeeling = true;
+      }
+    }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp]: Can we create a function in CPUUtils, and reuse it here? It makes codes easier to read.

**Review by hanhanW — CHANGES_REQUESTED** [review]

**Review by hanhanW — APPROVED** [review]

**Comment by banach-space** [comment]
@MaheshRavishankar Is this good to land?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 5]
```diff
@@ -211,6 +211,11 @@ getVectorPreProcStrategy(linalg::LinalgOp linalgOp) {
     return clPProcStrategy;
   }
 
+  // TODO: Implement heurystics for Convs
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Nit: type on heuristics.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 6]
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
...
+    auto peelAttrName = getPeelAttrName(ctx);
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.emplace_back(peelAttrName, peelAttrVal);
+    pipelineConfig = DictionaryAttr::get(ctx, attrs);
+  }
+
   return setOpConfigAndEntryPointFnTranslation(
       entryPointFn, convOp, tileSizes, scalableTileFlags,
-      DispatchLoweringPassPipeline::CPUConvTileAndDecomposeExpert);
+      DispatchLoweringPassPipeline::CPUConvTileAndDecomposeExpert, {}, {},
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Nit: please add `/*{argument_name}=*/` to the values that are set to default `{}`

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 7]
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
   return rootOperation;
 }
 
+StringAttr getPeelAttrName(MLIRContext *ctx) {
+  return StringAttr::get(ctx, "enable_loop_peeling");
+}
+
+bool isLoopPeelingEnabled(FunctionOpInterface *funcOp) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: This does not need to be a `FunctionOpInterface *`. Thats not the norm. Usually just pass these by value.

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 8]
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
   return rootOperation;
 }
 
+StringAttr getPeelAttrName(MLIRContext *ctx) {
+  return StringAttr::get(ctx, "enable_loop_peeling");
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: Nit: Define a static char array with the name an use that.
```
static const char[] kLoopPeeling = "enable_loop_peeling";
```
and use `kLoopPeeling` here.

**Review by MaheshRavishankar — APPROVED** [review]
I am unblocking, but really would be good to wait for Hanhan's comments.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 9]
```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
...
+  DictionaryAttr pipelineConfig;
+  if (vecPreProcStrategy == VectorPreProcStrategy::Peeling) {
+    // Enable peeling. To this end, attach extra info to the pipeline config.
+    // This will later be extracted by LLVMCPULowerExecutableTargetPass.
+    auto ctx = convOp.getContext();
+    SmallVector<NamedAttribute> attrs;
+    auto peelAttrName = getPeelAttrName(ctx);
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.emplace_back(peelAttrName, peelAttrVal);
+    pipelineConfig = DictionaryAttr::get(ctx, attrs);
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: nit: I think this could be clearer with some naming tweaks + a few less steps:
```c++
    auto context = convOp.getContext();
    auto enableLoopPeelingAttrName = getEnableLoopPeelingAttrName(context);
    auto trueValue = BoolAttr::get(context, true);
    pipelineConfig = DictionaryAttr::get(
        context,
        ArrayRef<NamedAttribute>({enableLoopPeelingAttrName, trueValue}));
```

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 10]
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
   return rootOperation;
 }
 
+StringAttr getPeelAttrName(MLIRContext *ctx) {
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: nit: This is `getEnableLoopPeelingAttrName()`

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 11]
```diff
@@ -63,4 +64,25 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
...
+
+  IREE::Codegen::TranslationInfoAttr translationInfo =
+      getTranslationInfo(*funcOp);
+  DictionaryAttr config = translationInfo.getConfiguration();
+  if (config) {
+    auto peelAttr = translationInfo.getConfiguration().getNamed(peelAttrName);
+    if (peelAttr && peelAttr->getValue() == trueAttr) {
+      return true;
+    }
+  }
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: nit: early exists (less nesting)
```c++
if (!config)
   return false;
```

**Review by MacDue — COMMENTED** [review]
a few nits:

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp (hunk 11 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: (similar naming nits here this is `enableLoopPeelingAttrName`)

Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: That's much cleaner, thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 9 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Thanks, this is much nicer! I thought that I needed a `SmallVector` here, but this makes me realise that we only need storage for the actual  contents.

**Comment by banach-space** [comment]
> I am unblocking, but really would be good to wait for Hanhan's comments.

Hanhan has already approved :) If there are no new comments and Ben approves, I will land this tomorrow. Thank you all!

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 12]
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
...
+
+  IREE::Codegen::TranslationInfoAttr translationInfo =
+      getTranslationInfo(funcOp);
+  DictionaryAttr config = translationInfo.getConfiguration();
+
+  if (config) {
+    return false;
+  }
+
+  auto peelAttr = translationInfo.getConfiguration().getNamed(peelAttrName);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: nit: Use config directly, i.e., `config.getNamed(...)`.

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 13]
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
   return rootOperation;
 }
 
+StringAttr getEnableLoopPeelingAttrName(MLIRContext *ctx) {
+  return StringAttr::get(ctx, kLoopPeelingAttrName);
+}
+
+bool isLoopPeelingEnabled(FunctionOpInterface funcOp) {
+  auto peelAttrName = getEnableLoopPeelingAttrName(funcOp->getContext());
+  auto trueAttr = BoolAttr::get(funcOp->getContext(), true);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: Do we use `UnitAttr` instead? The attribute name is `enable_loop_peeling`. It is clear that we want peeling when the attribute is present. What do you think?

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 14]
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
...
+
+  IREE::Codegen::TranslationInfoAttr translationInfo =
+      getTranslationInfo(funcOp);
+  DictionaryAttr config = translationInfo.getConfiguration();
+
+  if (config) {
+    return false;
+  }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: Should this be `!config`?

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp (hunk 13 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: +1 to a unit attribute

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 15]
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
...
+
+  IREE::Codegen::TranslationInfoAttr translationInfo =
+      getTranslationInfo(funcOp);
+  DictionaryAttr config = translationInfo.getConfiguration();
+
+  if (!config) {
+    return false;
+  }
+
+  auto peelAttr = translationInfo.getConfiguration().getNamed(peelAttrName);
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: nit: `peelAttr` -> `enableLoopPeelingAttr` 
nit:  `enableLoopPeelingAttrName`

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 16]
```diff
@@ -63,4 +66,28 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
...
+  if (peelAttr && peelAttr->getValue() == trueAttr) {
+    return true;
+  }
+
+  return false;
+}
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: Not relevant if using a `UnitAttr`, but `BoolAttr::getValue()` returns a normal `bool` (there's no need to construct a `trueAttr` attribute). Also, the `if` is redundant here. 
```suggestion
  return peelAttr && peelAttr->getValue();
}
```

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp (hunk 14 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: Yeah, realised that after pushing.

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 17]
```diff
@@ -63,4 +66,22 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
...
+
+  DictionaryAttr config = getTranslationInfo(funcOp).getConfiguration();
+
+  if (!config) {
+    return false;
+  }
+
+  auto enableLoopPeelingAttrName = getEnableLoopPeelingAttrName(context);
+  auto enableLoopPeelingAttr = config.getNamed(enableLoopPeelingAttrName);
+  return enableLoopPeelingAttr.has_value() ? true : false;
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: nit: There's no need to check a unit attribute (simply having the attribute enables peeling). Also, you don't need to construct a `StringAttr` to check (`.contains` can take a `StringRef`).
```suggestion
  return config.contains(kLoopPeelingAttrName);
```

**Review by MacDue — APPROVED** [review]
Final nit, otherwise LGTM.

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp (hunk 17 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: nit I just noticed: `kLoopPeelingAttrName` -> `kEnableLoopPeelingAttrName`

Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: Or or even just:
```suggestion
  return config && config.contains(kLoopPeelingAttrName);
```

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: (no need for the `auto context = funcOp->getContext();` too)

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp [hunk 18]
```diff
@@ -63,4 +66,18 @@ FailureOr<Operation *> getRootOperation(ArrayRef<Operation *> computeOps) {
...
+}
+
+bool isLoopPeelingEnabled(FunctionOpInterface funcOp) {
+  DictionaryAttr config = getTranslationInfo(funcOp).getConfiguration();
+
+  if (!config) {
+    return false;
+  }
+
+  return config.contains(kLoopPeelingAttrName);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: nit: I think we can collapse it to `return config && config.contains(...);`. What do you think?

**Review by hanhanW — APPROVED** [review]
Thanks for addressing all the comments! Just a final nit, please take a look.

File: compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp (hunk 18 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp]: Ben suggested exactly that and in my head I've already added that 🤦🏻 But that was imaginary code :) Thanks for pointing this out, sending update shortly.
